src/m365/entra/commands/app/app-permission-list.ts

```diff
@@ -0,0 +1,264 @@
+          return {
+            appId: resourceAccess.resourceAppId as string
+          };
+        });
+    const servicePrincipals = await Promise
+      .all(servicePrincipalsToResolve.map(servicePrincipalInfo =>
+        this.getServicePrincipal(servicePrincipalInfo, permissionType, logger) as ServicePrincipal));
+
+    const apiPermissions: ApiPermission[] = [];
+    (application.requiredResourceAccess as RequiredResourceAccess[]).forEach(requiredResourceAccess => {
+      const servicePrincipal = servicePrincipals
+        .find(servicePrincipal => servicePrincipal?.appId === requiredResourceAccess.resourceAppId as string);
+      const resourceName = servicePrincipal?.displayName as string ?? requiredResourceAccess.resourceAppId as string;
+      (requiredResourceAccess.resourceAccess as ResourceAccess[]).forEach(permission => {
+        if (permissionType === 'application' && permission.type === 'Scope') { return; }
+        if (permissionType === 'delegated' && permission.type === 'Role') { return; }
+
+        apiPermissions.push({
+          resource: resourceName,
+          resourceId: requiredResourceAccess.resourceAppId,
+          permission: this.getPermissionName(permission.id as string, permission.type as string, servicePrincipal),
+          type: permission.type === 'Role' ? 'Application' : 'Delegated'
+        });
+      });
+    });
+
+    return apiPermissions;
+  }
+
+  private getPermissionName(permissionId: string, permissionType: string, servicePrincipal: ServicePrincipal | undefined): string {
+    if (!servicePrincipal) {
+      return permissionId;
+    }
+
+    switch (permissionType) {
+      case 'Role':
+        return (servicePrincipal.appRoles as AppRole[])
+          .find(appRole => appRole.id === permissionId)?.value as string ?? permissionId;
+      case 'Scope':
+        return (servicePrincipal.oauth2PermissionScopes as PermissionScope[])
+          .find(permissionScope => permissionScope.id === permissionId)?.value as string ?? permissionId;
+    }
+    /* c8 ignore next 4 */
+    // permissionType is either 'Scope' or 'Role' but we need a safe default
+    // to avoid building errors. This code will never be reached.
+    return permissionId;
+  }
+
+  private async getServicePrincipal(servicePrincipalInfo: ServicePrincipalInfo, permissionType: string, logger: Logger): Promise<ServicePrincipal | undefined> {
+    if (this.verbose) {
+      await logger.logToStderr(`Retrieving service principal ${servicePrincipalInfo.appId}`);
+    }
+
+    const requestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/servicePrincipals?$filter=appId eq '${servicePrincipalInfo.appId}'&$select=appId,id,displayName`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+
+    const response = await request.get<{ value: ServicePrincipal[] } | ServicePrincipal>(requestOptions);
+
+    if ((servicePrincipalInfo.id) ||
```

Comment:
sorry I am a bit lost 🙂.
Is it possible to have `(servicePrincipalInfo.id)` here? From what I understand we will have only `appId` right? since this is the only property we map in line 148